org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ImportRootProjectOperation.java

```diff
@@ -0,0 +1,122 @@
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.SubMonitor;
+
+import org.eclipse.buildship.core.CorePlugin;
+import org.eclipse.buildship.core.UnsupportedConfigurationException;
+import org.eclipse.buildship.core.configuration.BuildConfiguration;
+import org.eclipse.buildship.core.configuration.GradleProjectNature;
+import org.eclipse.buildship.core.configuration.ProjectConfiguration;
+import org.eclipse.buildship.core.workspace.NewProjectHandler;
+import org.eclipse.buildship.core.workspace.WorkspaceOperations;
+
+/**
+ * Imports the root project into the workspace.
+ *
+ * @author Donat Csikos
+ */
+public final class ImportRootProjectOperation {
+
+    private final BuildConfiguration buildConfiguration;
+    private final NewProjectHandler newProjectHandler;
+
+    public ImportRootProjectOperation(BuildConfiguration buildConfiguration, NewProjectHandler newProjectHandler) {
+        this.buildConfiguration = Preconditions.checkNotNull(buildConfiguration);
+        this.newProjectHandler = Preconditions.checkNotNull(newProjectHandler);
+    }
+
+    public void run(IProgressMonitor monitor) {
+        try {
+            runInWorkspace(monitor);
+        } catch (Exception e) {
+            throw new ImportRootProjectException(e);
+        }
+    }
+
+    private void runInWorkspace(IProgressMonitor monitor) throws CoreException {
+        ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
+
+            @Override
+            public void run(IProgressMonitor monitor) throws CoreException {
+                SubMonitor progress = SubMonitor.convert(monitor);
+                progress.setTaskName("Importing root project");
+                progress.setWorkRemaining(3);
+
+                File rootDir = ImportRootProjectOperation.this.buildConfiguration.getRootProjectDirectory();
+                verifyNoWorkspaceRootIsImported(rootDir, progress.newChild(1));
+                saveProjectConfiguration(ImportRootProjectOperation.this.buildConfiguration, rootDir, progress.newChild(1));
+                importRootProject(rootDir, progress.newChild(1));
+            }
+        }, monitor);
+    }
+
+    private void verifyNoWorkspaceRootIsImported(File rootDir, IProgressMonitor monitor) {
+        File workspaceRoot = ResourcesPlugin.getWorkspace().getRoot().getLocation().toFile();
+        if (rootDir.equals(workspaceRoot)) {
+            throw new UnsupportedConfigurationException(String.format("Project %s location matches workspace root %s", rootDir.getName(), workspaceRoot.getAbsolutePath()));
+        }
+    }
+
+    private void saveProjectConfiguration(BuildConfiguration buildConfiguration, File rootDir, IProgressMonitor monitor) {
+        ProjectConfiguration projectConfiguration = CorePlugin.configurationManager().createProjectConfiguration(buildConfiguration, rootDir);
```

Comment:
❌ Why do we need to duplicate this here? Shouldn't this be done by the synchronization action that comes next?

Reply:
The user could use the "Add Gradle Nature" action instead, but I get what you mean.